src/shaka_scheme/system/base/Data.cpp

```diff
@@ -63,15 +71,125 @@ shaka::Data::~Data() {
     this->boolean.~Boolean();
     break;
   }
-//  case shaka::Data::Type::DATA_NODE: {
-//    this->data_node::~data_node();
-//    break;
-//  }
+  case shaka::Data::Type::DATA_PAIR: {
+    this->data_pair.~DataPair();
+    break;
+  }
 //  case shaka::Data::Type::ENVIRONMENT: {
 //    this->environment::~environment();
 //    break;
 //  }
+  case shaka::Data::Type::NULL_LIST: {
+    break;
+  }
+  default: {}
   }
 
 }
 
+template<>
+shaka::String& shaka::Data::get<shaka::String>() {
+  if (this->get_type() != Type::STRING) {
+    throw new shaka::TypeException(3, "Could not get() String from Data");
+  }
+  return this->string;
+}
+
+template<>
+shaka::Symbol& shaka::Data::get<shaka::Symbol>() {
+  if (this->get_type() != Type::SYMBOL) {
+    throw new shaka::TypeException(4, "Could not get() Symbol from Data");
+  }
+  return this->symbol;
+}
+
+template<>
+shaka::Boolean& shaka::Data::get<shaka::Boolean>() {
+  if (this->get_type() != Type::BOOLEAN) {
+    throw new shaka::TypeException(5, "Could not get() Boolean from Data");
+  }
+  return this->boolean;
+}
+

```

Comment:
"Please review whether it is necessary to use references to allow ourselves to directly set the car or cdr for a DataPair if we're trying to get a DataPair from a Data (we don't want to modify a copy -- we want to modify the actual data!)"

I believe this is necessary as things are currently implemented, having looked at the implementations of set_car and set_cdr , where you are basically assigning pointer variables. If you want the pointers representing the car and cdr in the _original DataPair_ retrieved via a call to `shaka::Data::get` to be changed, it is necessary to return a reference to this DataPair. Otherwise, we will just be changing what the pointer fields of a copy of that DataPair object point to, and this does not obey the proper semantics of set-car! and set-cdr! in scheme.

Example: Evaluated in chez scheme REPL

`(define lst '((1 2) 3 4 5 6))
(define car-lst (car lst))
car-lst => '(1 2)
(set-car! car-lst 2)
car-lst => '(2 2)
lst => '((2 2) 3 4 5 6)`

As you can see, we mutated the original car of the list bound to the `lst` symbol in the environment indirectly, by calling set-car! on the binding of that car to the symbol `lst-car`. Thus, returning a reference from the call to get is necessary in the case of returning a DataPair. However, I might caution against returning a reference in the case where what we are "getting" out of a Data object is not a DataPair.

Example: Evaluated in chez scheme REPL

`(define lst '(1 2 3 4 5))
(define car-lst (car lst))
car-lst => 1
(set! car-lst 5)
car-lst => 5
lst => '(1 2 3 4 5)`

In the case of atomic contents, we cannot indirectly modify the contents of a list via calls to `set!` on a symbol that is bound to them. In light of this, there is a chance that returning a reference to the actual data in all cases might break these semantics, but further testing is likely required to confirm/deny this. Otherwise, the implementation looks rock-solid to me.

Reply:
I think the current way that it's being implemented, it's a shallow copy for going from `Data -> Data`. But for primitive data, a shallow copy means actually copying the actual stuff.

Though, I do agree that we should have a unit test for this.